service/history/decision/handler.go

```diff
@@ -841,6 +843,25 @@ func failDecisionHelper(
 		return nil, err
 	}
 
+	domainName := domainEntry.GetInfo().Name
+	maxAttempts := handler.config.DecisionRetryMaxAttempts(domainName)
+	if maxAttempts > 0 && mutableState.GetExecutionInfo().DecisionAttempt > int64(maxAttempts) {
+		executionInfo := mutableState.GetExecutionInfo()
+		handler.logger.Error("Decision attempt exceeds limit.",
+			tag.WorkflowDomainID(executionInfo.DomainID),
+			tag.WorkflowID(executionInfo.WorkflowID),
+			tag.WorkflowRunID(executionInfo.RunID))
+
+		if _, err := mutableState.AddWorkflowExecutionTerminatedEvent(
+			mutableState.GetNextEventID(),
+			common.FailureReasonDecisionAttemptsExceedsLimit,
+			[]byte("Decision attempt exceeded limit."),
```

Comment:
nit: shall we use `cause` and/or `details` parameter here?

Reply:
These two are in the DecisionTaskFailedEvent already, so I don't think it's very useful to add them termination event, too.

Reply:
hmmm, in most cases yes. But if we are unlucky and the first decision failure is due to decision timeout then later failed events and reason won't be recorded in history I think. 

A similar issue is when using the `--non-deterministic-only` flag when doing a batch reset. Most workflow's last event is decisionTaskFailed event which contains the detailed info for the non-deterministic error. But a small number of them are end by decisionTaskTimeout event, which can't be caught by the `--non-deterministic-only` flag.